kusto/ingest/result.go

```diff
@@ -0,0 +1,172 @@
+	// If not checking status, just return queued
+	if !r.reportToTable {
+		r.record.Status = Queued
+		return r
+	}
+
+	// Get table URI
+	resources, err := mgr.Resources()
+	if err != nil {
+		r.record.Status = StatusRetrievalFailed
+		r.record.FailureStatus = Permanent
+		r.record.Details = "Failed getting status table URI: " + err.Error()
+		return r
+	}
+
+	if len(resources.Tables) == 0 {
+		r.record.Status = StatusRetrievalFailed
+		r.record.FailureStatus = Permanent
+		r.record.Details = "Ingestion resources do not include a status table URI: " + err.Error()
+		return r
+	}
+
+	// create a table client
+	client, err := status.NewTableClient(*resources.Tables[0])
+	if err != nil {
+		r.record.Status = StatusRetrievalFailed
+		r.record.FailureStatus = Permanent
+		r.record.Details = "Failed Creating a Status Table client: " + err.Error()
+		return r
+	}
+
+	// Write initial record
+	r.record.Status = Pending
+	recordMap := r.record.ToMap()
+	err = client.Write(r.record.IngestionSourceID.String(), recordMap)
+	if err != nil {
+		r.record.Status = StatusRetrievalFailed
+		r.record.FailureStatus = Permanent
+		r.record.Details = "Failed writing initial status record: " + err.Error()
+		return r
+	}
+
+	r.tableClient = client
+	return r
+}
+
+// Wait returns a channel that can be checked for ingestion results.
+// In order to check actual status please use the IngestionStatus option when ingesting data.
+func (r *Result) Wait(ctx context.Context) chan error {
+	ch := make(chan error, 1)
+
+	go func() {
+		defer close(ch)
+
+		if !r.record.Status.IsFinal() && r.reportToTable == true {
+			r.poll(ctx)
+		}
+
+		if !r.record.Status.IsSuccess() {
+			ch <- r.record
+		}
+	}()
+
+	return ch
+}
+
+func (r *Result) poll(ctx context.Context) {
+	const pollInterval = 10 * time.Second
+
+	// create a table client
+	if r.tableClient != nil {
+		// Create a ticker to poll the table in 10 second intervals.
+		timer := time.NewTimer(pollInterval)
+		defer timer.Stop()
+
+		for {
+			select {
+			// In case the user canceled the wait, return current known state.
+			case <-ctx.Done():
+				// return a canceld state.
+				r.record.Status = StatusRetrievalCanceled
+				r.record.FailureStatus = Transient
+				return
+
+			// Whenever the ticker fires.
+			case <-timer.C:
+				// read the current state
+				smap, err := r.tableClient.Read(r.record.IngestionSourceID.String())
+				if err != nil {
+					// Read failure
+					r.record.Status = StatusRetrievalFailed
+					r.record.FailureStatus = Transient
+					r.record.Details = "Failed reading from Status Table: " + err.Error()
+					return
+				}
+
+				// convert the data into a record and send it if the state is final.
+				r.record.FromMap(smap)
+				if r.record.Status.IsFinal() {
+					return
+				}
+
+				timer.Reset(pollInterval)
+			}
+		}
+	}
+}
+
+// GetIngestionStatus extracts the ingestion status code from an ingestion error
+func GetIngestionStatus(err error) (StatusCode, error) {

```

Comment:
I think we should combine this and GetIngestionFailureStatus() and remove Get*, and drop the extra error, aka:

```go
// IngestionStatus retrieves the StatusCode and FailureStatusCode if the error contains them.  It will return
// UnknownStatus and UnknownFailure if the error does not contain them.
fun IngestionStatus(err error) (StatusCode, FailureStatusCode) {
        if s, ok := err.(statusRecord); ok {
		return s.Status, s.FailureStatus
	}
        return UnknownStatus, UnknownFailure
}
```

This change would require adding: UnknownStatus and changing Unknown to UnknownFailure.

Reply:
I'm bouncing around with this, why we diff on this.  I get your point of view, at least on the extra "error".  But it seems unclean, and I know I haven't seen that.

But maybe my memory is faulty.  I wanted prior art before I commented back.  So I went digging in the standard library, but wasn't finding anything (they don't really deal a lot with status codes via errors, they handle it with explicit errors).  

I think maybe I got this from stubby, so went and looked at gRPC status calls and found: 
https://github.com/grpc/grpc-go/blob/master/status/status.go#L100

This code base is doing what we are doing (except we split status and failure instead of just a single code).  They return "error" that can later be converted to its *Status type if wanted (we don't do this, yet) or have its Code retrieved.  

So, let's sit and chat about what we should do here.  I'd like to talk the pro's and cons, as I don't feel great about either one.